File: C++_version/FCFS.cpp
#include "FCFS.hpp"
#include <iostream>

FCFS::FCFS(std::vector<Process> &process) : Scheduler(process) {};
FCFS::FCFS() {};

int FCFS::schedule(){
	int first = 9999;
	int i;
	for(i=0;i<process_queue.size();i++){
		if(process_queue[i]<9999)
			first=process_queue[i].getID();
	}
	running_id = first;
	process_queue[i].setID(process_queue[i].getBurst_time()-1);
	if(process_queue[i].getBurst_time() == 0)
		process_queue.remove(i);
}

File: java_version/FCFS.java
import java.util.*;

public class FCFS extends Scheduler {

    public FCFS(List<Process> processes) {
        super(processes);
    }

    public FCFS() {
        super();
    }

    @Override
    public int schedule() {
        // Sort processes by arrival time (FCFS)
        processQueue.sort(Comparator.comparingInt(Process::getArrivalTime));

        while (!isFinished()) {
            // Find next process that has arrived and is not finished
            for (Process process : processQueue) {
                if (process.getArrivalTime() <= currentTime && process.getExecutionTime() < process.getBurstTime()) {
                    runningProcess = process;
                    break;
                }
            }

            if (runningProcess != null) {
                // Execute the running process
                runningProcess.setExecutionTime(runningProcess.getExecutionTime() + 1);

                if (runningProcess.getExecutionTime() == runningProcess.getBurstTime()) {
                    // Process finished: calculate turnaround and waiting times
                    int turnaroundTime = currentTime + 1 - runningProcess.getArrivalTime();
                    int waitingTime = turnaroundTime - runningProcess.getBurstTime();

                    runningProcess.setTurnaroundTime(turnaroundTime);
                    runningProcess.setWaitingTime(waitingTime);

                    runningProcess = null; // No running process anymore
                }
            }

            currentTime++; // Move time forward
        }

        return currentTime; // Return total time taken
    }
}
